openpype/modules/sync_server/sync_server.py

```diff
@@ -189,6 +205,166 @@ def _site_is_working(module, project_name, site_name, site_config):
+    anatomy: Anatomy = None,
+    asset_doc: dict = None,
+) -> str:
+    """Download the last published workfile, and return its path.
+
+    Args:
+        host_name (str): Host name.
+        project_name (str): Project name.
+        asset_name (str): Asset name.
+        task_name (str): Task name.
+        last_published_workfile_path (str): Last published workfile
+            path.
+        workfile_representation (dict): Workfile representation.
+        subset_id (str): Subset ID.
+        last_version_doc (dict): Last version doc.
+        anatomy (Anatomy, optional): Anatomy (Used for optimization).
+            Defaults to None.
+        asset_doc (dict, optional): Asset doc (Used for optimization).
+            Defaults to None.
+
+    Returns:
+        str: New local workfile path.
+    """
+
+    if not anatomy:
+        anatomy = Anatomy(project_name)
+
+    if not asset_doc:
+        asset_doc = get_asset_by_name(project_name, asset_name)
+
+    # Get sync server module
+    sync_server = ModulesManager().modules_by_name.get("sync_server")
+    if not sync_server or not sync_server.enabled:
+        print("Sync server module is disabled or unavailable.")
+        return
+
+    if subset_id is None:
+        print(
+            "Not any matched subset for task '{}' of '{}'.".format(
+                task_name, asset_name
+            )
+        )
+        return
+
+    if not workfile_representation:
+        print(
+            "Not published workfile for task '{}' and host '{}'.".format(
+                task_name, host_name
+            )
+        )
+        return
+
+    # If representation isn't available on remote site, then return.
+    if not sync_server.is_representation_on_site(
+        project_name,
+        workfile_representation["_id"],
+        sync_server.get_remote_site(project_name),
+    ):
+        print(
+            "Representation for task '{}' and host '{}'".format(
+                task_name, host_name
+            )
+        )
+        return
+
+    # Get local site
+    local_site_id = get_local_site_id()
+
+    # Add workfile representation to local site
+    representation_ids = {workfile_representation["_id"]}
+    representation_ids.update(
+        get_linked_representation_id(
+            project_name, repre_id=workfile_representation["_id"]
+        )
+    )
+    for repre_id in representation_ids:
+        if not sync_server.is_representation_on_site(project_name, repre_id,
+                                                     local_site_id):
+            sync_server.add_site(
+                project_name,
+                repre_id,
+                local_site_id,
+                force=True,
+                priority=99
+            )
+    sync_server.reset_timer()
+
+    # While representation unavailable locally, wait.
+    while not sync_server.is_representation_on_site(
+        project_name, workfile_representation["_id"], local_site_id
+    ):
+        sleep(5)
```

Comment:
How heavy is the `is_representation_on_site` check? I'd expect that it might make much more sense to prior to this just get the "expected files list" from the sync server and periodically just check whether the file exists on disk (avoiding any database calls, etc.)? And then maybe after do a database check if needed? I'm asking because 5 seconds seems quite long to be an interval checking at if this is used in e.g. a prelaunch hook where the user is actively waiting for a result to happen.

Also, this while loop might need a maximum timeout too?

Reply:
I don't think we can reliably test presence of a file on the disk as some OSs will create a file there immediately and file changes until is fully downloaded. We could query DB for file size and compare it against file metadata from the disk. But can we get them reliably from all OSs and would it be actually faster? I don't feel that this would happen too often and it is quick check against indexed `_id`.

Was thinking about maximum timeout too, but what is reasonable value? 2 minutes, 5 minutes, 10?

Reply:
Added check for failed downloads, it could happen that downloading fails, so there is a limit how many tries should be attempted.